File: source/go/astar.go
package main

import "fmt"

const ROWS = 5
const COLS = 5

type Cell struct {
	r, c int
}

func absVal(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func heuristic(r1, c1, r2, c2 int) int {
	return absVal(r1-r2) + absVal(c1-c2)
}

func cellKey(r, c int) int {
	return r*COLS + c
}

func astar(grid [][]int, sr, sc, er, ec int) []Cell {
	const INF = 999999999
	totalCells := ROWS * COLS

	gScore := make([]int, totalCells)
	fScore := make([]int, totalCells)
	cameFrom := make([]int, totalCells)
	closed := make([]bool, totalCells)
	for i := 0; i < totalCells; i++ {
		gScore[i] = INF
		fScore[i] = INF
		cameFrom[i] = -1
	}

	startKey := cellKey(sr, sc)
	gScore[startKey] = 0
	fScore[startKey] = heuristic(sr, sc, er, ec)

	openList := []int{startKey}

	dr := []int{-1, 1, 0, 0}
	dc := []int{0, 0, -1, 1}

	for len(openList) > 0 {
		bestIdx := 0
		bestF := fScore[openList[0]]
		for i := 1; i < len(openList); i++ {
			if fScore[openList[i]] < bestF {
				bestF = fScore[openList[i]]
				bestIdx = i
			}
		}

		current := openList[bestIdx]
		openList = append(openList[:bestIdx], openList[bestIdx+1:]...)

		cr := current / COLS
		cc := current % COLS

		if cr == er && cc == ec {
			var path []Cell
			c := current
			for c != -1 {
				pr := c / COLS
				pc := c % COLS
				path = append([]Cell{{pr, pc}}, path...)
				c = cameFrom[c]
			}
			return path
		}

		closed[current] = true

		for d := 0; d < 4; d++ {
			nr := cr + dr[d]
			nc := cc + dc[d]
			if nr < 0 || nr >= ROWS || nc < 0 || nc >= COLS {
				continue
			}
			if grid[nr][nc] == 1 {
				continue
			}
			nkey := cellKey(nr, nc)
			if closed[nkey] {
				continue
			}
			tentativeG := gScore[current] + 1
			if tentativeG < gScore[nkey] {
				cameFrom[nkey] = current
				gScore[nkey] = tentativeG
				fScore[nkey] = tentativeG + heuristic(nr, nc, er, ec)
				found := false
				for _, k := range openList {
					if k == nkey {
						found = true
						break
					}
				}
				if !found {
					openList = append(openList, nkey)
				}
			}
		}
	}
	return nil
}

func main() {
	grid := [][]int{
		{0, 0, 0, 0, 0},
		{0, 1, 1, 0, 0},
		{0, 0, 1, 0, 0},
		{0, 0, 0, 1, 0},
		{0, 0, 0, 0, 0},
	}

	path := astar(grid, 0, 0, 4, 4)
	fmt.Printf("path length: %d\n", len(path))
	coords := ""
	for i, p := range path {
		if i > 0 {
			coords += " "
		}
		coords += fmt.Sprintf("(%d,%d)", p.r, p.c)
	}
	fmt.Printf("path: %s\n", coords)
}

File: source/python/astar.py
ROWS = 5
COLS = 5

grid = [
    [0, 0, 0, 0, 0],
    [0, 1, 1, 0, 0],
    [0, 0, 1, 0, 0],
    [0, 0, 0, 1, 0],
    [0, 0, 0, 0, 0],
]

def heuristic(r1, c1, r2, c2):
    return abs(r1 - r2) + abs(c1 - c2)

def cell_key(r, c):
    return r * COLS + c

def astar(grid, sr, sc, er, ec):
    INF = 999999999
    total_cells = ROWS * COLS
    g_score = [INF] * total_cells
    f_score = [INF] * total_cells
    came_from = [-1] * total_cells
    closed = [False] * total_cells

    start_key = cell_key(sr, sc)
    g_score[start_key] = 0
    f_score[start_key] = heuristic(sr, sc, er, ec)
    open_list = [start_key]

    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]

    while open_list:
        best_idx = 0
        best_f = f_score[open_list[0]]
        for i in range(1, len(open_list)):
            if f_score[open_list[i]] < best_f:
                best_f = f_score[open_list[i]]
                best_idx = i

        current = open_list.pop(best_idx)
        cr = current // COLS
        cc = current % COLS

        if cr == er and cc == ec:
            path = []
            c = current
            while c != -1:
                pr = c // COLS
                pc = c % COLS
                path.insert(0, {"r": pr, "c": pc})
                c = came_from[c]
            return path

        closed[current] = True

        for d in range(4):
            nr = cr + dr[d]
            nc = cc + dc[d]
            if nr < 0 or nr >= ROWS or nc < 0 or nc >= COLS:
                continue
            if grid[nr][nc] == 1:
                continue
            nkey = cell_key(nr, nc)
            if closed[nkey]:
                continue
            tentative_g = g_score[current] + 1
            if tentative_g < g_score[nkey]:
                came_from[nkey] = current
                g_score[nkey] = tentative_g
                f_score[nkey] = tentative_g + heuristic(nr, nc, er, ec)
                if nkey not in open_list:
                    open_list.append(nkey)

    return []

path = astar(grid, 0, 0, 4, 4)
print(f"path length: {len(path)}")
coords = " ".join(f"({p['r']},{p['c']})" for p in path)
print(f"path: {coords}")
